In [1]:
# Importación de paquetes
import numpy as np # Cálculo numérico
import pandas as pd # Análisis de datos

### Parámetros del Algoritmo

In [2]:
# Tamaño de la población
numero_individuos = 20

# Constantes de la función de Ackley
a = 20
b = 0.02
c = 2 * np.pi

# Número de corridas
total_epocas = 10

# Probabilidad de mutación
m = 0.45

### Función Fitness

In [3]:
# función de Ackley 1
def fitness (fenotipo):
    d = len(fenotipo) # número de dimensiones
    sum1 = 0 # suma dentro de la primera función exp
    sum2 = 0 # suma dentro de la segunda función exp
    for i in range(0, d - 1): # cálculo de las sumas
        sum1 += fenotipo[i]**2
        sum2 += np.cos(c * fenotipo[i])
    return -a * np.exp(-b * np.sqrt((1/d) * sum1)) - np.exp((1/d) * sum2) + a + np.exp(1)

### Funciones de Codificación Genética

In [4]:
def decodificar (genotipo):
    fenotipo = np.zeros(3) # Inicializar variable
    coordenada = 2 # Coordenada actual = variable 3
    
    # Creamos un conjunto de índices que van desde 59 hasta 0 para representar la posición de los genes en el cromosoma
    posiciones = np.arange(len(genotipo) - 1, -1, -1)

    # Expresión decimal de la cadena binaria que representa una variable dentro del cromosoma
    decimal = 0.0

    # Valor real decodificado
    x = 0.0

    for posicion in posiciones:
        # Si el resto de la división por 20 es mayor que cero, estamos en un gen dentro de la misma variable.
        exponente = posicion % 20
        if exponente > 0:
            decimal += genotipo[posicion] * 2**(exponente)
            continue

        # Cuando llegue a cero, terminamos de decodificar la variable y cambiamos de coordenada
        decimal += genotipo[posicion]
        x = -35 + decimal * (70 / (2**20 - 1))
        fenotipo[coordenada] = np.round(x, 4)
        coordenada -= 1
        decimal = 0.0
    
    return fenotipo

# Función auxiliar para verificar que un individuo representa una solución factible
def validar (cromosoma):
    factible = True
    fenotipo = decodificar(cromosoma)

    # Para marcar un cromosoma como no factible, basta con que una de las coordenadas caiga fuera del rango [-32, 32]
    for i in range(0, len(fenotipo)):
        if fenotipo[i] > 35 or fenotipo[i] < -35:
            factible = False
            break

    return factible

# Función auxiliar para visualizar un cromosoma como una cadena de caracteres
def cromosoma_to_string (cromosoma):
    cadena = ''
    for gen in cromosoma:
        cadena = f'{cadena}{gen}'
    return cadena

### Función para Generar un Individuo Factible

In [5]:
def generar_individuo ():
    # Creamos un cromosoma genérico de 60 genes todos con alelo = 0
    cromosoma = np.zeros(60, int)

    # El objetivo es ir añadiendo 1's de menor a mayor posición según una ley Uniforme(0, 1) a cada coordenada hasta que se supere el máximo valor permitido: 32. Esto para que las soluciones que se generen de forma aleatoria estén dentro del dominio

    posiciones = np.arange(59, -1, -1)

    for posicion in posiciones:
        u = np.random.random()
        alelo = 1 if u < 0.5 else 0 # 50% de probabilidad
        
        # Si el alelo obtenido aleatoriamente es 0, no hay que hacer nada ya que la posición del cromosoma se encuentra en 0 por defecto. Si no, asignamos a la posición el alelo 1 y validamos que sea un cromosoma válido, si no es válido, dejamos el gen en 0
        if alelo == 1:
            cromosoma[posicion] = 1
            if not validar(cromosoma):
                cromosoma[posicion] = 0

    return cromosoma

### Funciones de Cruce y Mutación

La recombinación se hará copiando los genes del primer padre desde el gen 0 al gen 29 para luego copiar los genes del segundo padre desde el gen 30 hasta el gen 59 (mitad de los genes del padre 1 y mitad de los genes del padre 2)

In [6]:
def cruzar (padre1, padre2):
    nuevo_individuo = np.zeros(60, dtype=int) # Creamos una cadena binaria genérica
    for gen in range(0, 60):
        alelo = padre1[gen] if gen < 29 else padre2[gen]
        nuevo_individuo[gen] = alelo
    return nuevo_individuo

La mutación consistirá simplemente en invertir el alelo de cada gen:

In [7]:
def mutar (individuo):
    individuo_mutado = np.zeros(len(individuo), dtype=int)
    for gen in range(0, len(individuo)):
        individuo_mutado[gen] = 1 if individuo[gen] == 0 else 0
    return individuo_mutado

### Generación de la población inicial

In [8]:
# Inicialización de la tabla para registrar y mostrar la población
poblacion = []

# Añadimos a cada fila un individuo junto con su valor de fitness
for i in range(0, numero_individuos):
    genotipo = generar_individuo()
    fenotipo = decodificar(genotipo)
    poblacion.append({
        'Individuo' : cromosoma_to_string(genotipo),
        'Fenotipo' : fenotipo,
        'Fitness': fitness(fenotipo)
    })

# Creación del dataframe para visualizar la población
df_poblacion = pd.DataFrame(poblacion)

# Visualización de la población
poblacion_inicial = df_poblacion.style.set_properties(**{
    'font-size' : '12pt'
})
poblacion_inicial

### Fitness medio de la Población Inicial

In [9]:
fitness_medio_inicial = df_poblacion.Fitness.mean()
print(fitness_medio_inicial)

6.223763021358517


El objetivo es obtener una población (conjunto de soluciones) que tengan un mejor desempeño, es decir, que su fitness medio sea menor que 6,2237. Se realizará un total de 10 corridas (épocas)

### Algoritmo Genético

El mecanismo de selección de los dos padres será aquellos con mejor fitness. Dado que estamos minimizando, se trata de los dos que tengan los menores valores en la función de Ackley 2. El peor individuo de la población será el que tenga mayor valor en la función de Ackley 2.

In [10]:
# Vamos a registrar cada nuevo individuo con un nuevo índice para saber al final cuáles soluciones fueron añadidas y cuáles fueron descartadas
indice_individuo = len(df_poblacion)

# criterio de parada = número total de épocas alcanzadas
for _ in range(0, total_epocas):
    # Ordenamos la población según sus valores de fitness de forma creciente
    df_poblacion = df_poblacion.sort_values(by=['Fitness'])

    # Tomamos los dos mejores individuos, estos son los dos primeros en la lista de población ya que se ha ordenado de forma ascendente según su fitness
    padre1 = df_poblacion.Individuo.values[0]
    padre2 = df_poblacion.Individuo.values[1]

    # Los individuos en el dataframe se encuentran almacenadas como una cadena de caracteres, para poder operar con ellos con el paquete numpy, los convertimos a arreglos utilizando la función "fromiter" que itera sobre la cadena y devuelve el valor de cada caracter como una entrada del arreglo
    padre1 = np.fromiter(padre1, int)
    padre2 = np.fromiter(padre2, int)

    # Realizamos la recombinación para obtener un nuevo individuo
    nuevo_individuo = cruzar(padre1, padre2)

    # Con una probabilidad 'm' mutamos el nuevo cromosoma
    u = np.random.uniform(0, 1) # generamos un valor aleatorio entre 0 y 1 según una distribución uniforme
    if u < m:
        nuevo_individuo = mutar(nuevo_individuo)

    # Creamos un dataframe de una fila con el nuevo individuo con el fin de añadirlo a la población actual
    nuevo_fenotipo = decodificar(nuevo_individuo)
    nuevo_fitness = fitness(nuevo_individuo)
    df_nuevo_individuo = pd.DataFrame([[nuevo_fenotipo, cromosoma_to_string(nuevo_individuo), nuevo_fitness]], columns=['Fenotipo', 'Individuo', 'Fitness'], index=pd.Index([indice_individuo]))

    # Añadimos el nuevo individuo a la población, obteniéndose una nueva época con población de tamaño 21
    df_poblacion = pd.concat([df_poblacion, df_nuevo_individuo])

    # Volvemos a ordenar la población en orden creciente según el fitness
    df_poblacion = df_poblacion.sort_values(by=['Fitness'])

    # Eliminamos el último individuo de la población, el cual es el peor, ya que tiene el mayor valor de fitness
    df_poblacion = df_poblacion[:-1]

    indice_individuo += 1

### Población de la 10º Generación

In [11]:
poblacion_final = df_poblacion.style.set_properties(**{
    'font-size' : '12pt'
})
poblacion_final

### Fitness Medio de la Población Final

In [12]:
fitness_medio_final = df_poblacion.Fitness.mean()
print(fitness_medio_final)

2.424803407829007


### Conclusión

Luego de aplicar el AG al problema de minimización de la función de Ackley 1, hemos obtenido una mejora en el desempeño promedio de la población, de una media de 6,2237 a una media de 2.4248. Donde la solución con mejor desempeño es 

<center>[ 7,8007 -3,1406  29,6205 ]</center>

con un valor de la función de Ackley 1 de 0,316323; correspondiente al individuo Nº 24 definido por el siguiente cromosoma:

<center>110011100001001110010101110000010010111000011100101000110111</center>

nacido en la 6ta generación del algoritmo.